File: Q3_Cpp/src/Shipment/ShipmentCollection.cpp
#include "../../include/Shipment/ShipmentCollection.h"
#include <algorithm>

// Constructor for ShipmentCollection
ShipmentCollection::ShipmentCollection() {
    // The vector 'shipments' is automatically initialized
}

void ShipmentCollection::addShipment(const std::shared_ptr<Shipment>& shipment) {
    shipments.push_back(shipment);
}

std::vector<std::shared_ptr<Shipment>> ShipmentCollection::getAllShipments() const {
    return shipments; // Returning a copy of the shipments vector
}

std::shared_ptr<Shipment> ShipmentCollection::getShipmentById(const std::string& shipmentId) const {
    auto it = std::find_if(shipments.begin(), shipments.end(), 
        [&shipmentId](const std::shared_ptr<Shipment>& shipment) {
            return shipment->getId() == shipmentId;
        });

    if (it != shipments.end()) {
        return *it;
    }
    return nullptr; // If shipment is not found
}

bool ShipmentCollection::removeShipment(const std::string& shipmentId) {
    auto it = std::remove_if(shipments.begin(), shipments.end(), 
        [&shipmentId](const std::shared_ptr<Shipment>& shipment) {
            return shipment->getId() == shipmentId;
        });

    if (it != shipments.end()) {
        shipments.erase(it, shipments.end());
        return true;
    }
    return false; // If no shipment was removed
}

std::shared_ptr<Shipment> ShipmentCollection::getShipmentByCustomerId(const std::string& customerId) const {
    auto it = std::find_if(shipments.begin(), shipments.end(), 
        [&customerId](const std::shared_ptr<Shipment>& shipment) {
            return shipment->getCustomer()->getId() == customerId;
        });

    if (it != shipments.end()) {
        return *it;
    }
    return nullptr; // If shipment is not found
}

File: Q2/src/main/java/com/yourorganization/Shipment/ShipmentCollection.java
package com.yourorganization.Shipment; // Package for handling shipment collections

import java.util.ArrayList;
import java.util.List; // Imports for list functionality


public class ShipmentCollection {
    private List<Shipment> shipments; // List to store shipments

    // Constructor for ShipmentCollection
    public ShipmentCollection() {
        this.shipments = new ArrayList<>(); // Initializes the list of shipments
    }

    // Add a shipment to the collection
    public void addShipment(Shipment shipment) {
        shipments.add(shipment); // Appends a new shipment to the list
    }

    // Retrieve all shipments in the collection
    public List<Shipment> getAllShipments() {
        return new ArrayList<>(shipments); // Returns a copy of the shipments list
        // Once again a copy is returned not the original
    }

    // Find a shipment by its ID
    public Shipment getShipmentById(String shipmentId) {
        for (Shipment shipment : shipments) {
            if (shipment.getId().equals(shipmentId)) {
                return shipment; // Returns the matching shipment
            }
        }
        return null; // Return null if no match found
    }

    // Remove a shipment from the collection by ID
    public boolean removeShipment(String shipmentId) {
        return shipments.removeIf(shipment -> shipment.getId().equals(shipmentId)); // Removes the shipment
    }

    // Find a shipment by the customer ID
    public Shipment getShipmentByCustomerId(String customerId) {
        for (Shipment shipment : shipments) {
            if (shipment.getCustomer().getId().equals(customerId)) {
                return shipment; // Returns the shipment if customer ID matches
            }
        }
        return null;
    }
}
